packages/menu/src/MenuItem.ts

```diff
@@ -23,6 +23,67 @@ import { ActionButton } from '@spectrum-web-components/action-button';
 
 import menuItemStyles from './menu-item.css.js';
 import checkmarkStyles from '@spectrum-web-components/icon/src/spectrum-icon-checkmark.css.js';
+import { Menu } from './Menu.js';
+
+export class MenuItemRemovedEvent extends Event {
+    constructor() {
+        super('sp-menu-item-removed', {
+            bubbles: true,
+            composed: true,
+        });
+    }
+    get item(): MenuItem {
+        if (!this._item) {
+            this._item = this.composedPath()[0] as MenuItem;
+        }
+        return this._item;
+    }
+    _item?: MenuItem;
+    reset(): void {
+        this._item = undefined;
+    }
+}
+
+export class MenuItemAddedOrUpdatedEvent extends Event {
+    constructor() {
+        super('sp-menu-item-added', {
+            bubbles: true,
+            composed: true,
+        });
+    }
+    set focusRoot(root: Menu) {
+        this.item.menuData.focusRoot = this.item.menuData.focusRoot || root;
+    }
+    set selectionRoot(root: Menu) {
+        this.item.menuData.selectionRoot =
+            this.item.menuData.selectionRoot || root;
+    }
+    get item(): MenuItem {
+        if (!this._item) {
+            this._item = this.composedPath()[0] as MenuItem;
+        }
+        return this._item;
+    }
+    _item?: MenuItem;
+    set currentAncestorWithSelects(ancestor: Menu | undefined) {
+        this._currentAncestorWithSelects = ancestor;
+    }
+    get currentAncestorWithSelects(): Menu | undefined {
+        return this._currentAncestorWithSelects;
+    }
+    _currentAncestorWithSelects?: Menu;
+    reset(item: MenuItem): void {
+        this._item = undefined;
+        this._currentAncestorWithSelects = undefined;
+        item.menuData = {
+            focusRoot: undefined,
+            selectionRoot: undefined,
+        };
+    }
+}
+
+const addOrUpdateEvent = new MenuItemAddedOrUpdatedEvent();
+const removeEvent = new MenuItemRemovedEvent();
```

Comment:
I'm assuming this was a little performance improvement?

Reply:
Yeah, the `new`ing of a constructor (e.g. `new Event()`) can add up. Normally, we only do one or two at a time and it's not worth trying to optimize. However, in this case, we're causing ALL of the Menu Items to do this at once, so it seemed like something we could benefit from.